### PR TITLE
python3.pkgs.mmcv: init at 1.7.1

### PR DESCRIPTION
**Pull request by benxiao**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/mmcv/default.nix
Comment (fabaff) [inline pkgs/development/python-modules/mmcv/default.nix:33]:
```suggestion
    rev = "refs/tags/v${version}";
```

File: pkgs/development/python-modules/mmcv/default.nix
Comment (fabaff) [inline pkgs/development/python-modules/mmcv/default.nix:49]: Please add a comment while the tests are not executed.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (fabaff) [inline pkgs/development/python-modules/mmcv/default.nix:63]: Allows tools (e.g., `nix-update`) to add the changelog to the commit message while updating.

```suggestion
    homepage = "https://github.com/open-mmlab/mmcv";
    changelog = "https://github.com/open-mmlab/mmcv/releases/tag/v${version}";
```

File: pkgs/development/python-modules/mmcv/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/mmcv/default.nix:56]: suggestion: explain why pytestCheckHook cant be used

File: pkgs/development/python-modules/mmcv/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/mmcv/default.nix:87]: as per contribution guidelines, description should not start or contain name of the package

**Review by kirillrdy — COMMENTED** [review]
last commit message doesnt meet guidelines,

since this is an init PR, I would just squash all the commits

File: pkgs/development/python-modules/mmcv/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/mmcv/default.nix:9]: prefer camelCase https://nixos.org/manual/nixpkgs/stable/#sec-syntax

Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:10]: ok

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: have some of the tests running in the `checkPhase`.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: fixed

File: pkgs/development/python-modules/mmcv/default.nix
Comment (bcdarwin) [inline pkgs/development/python-modules/mmcv/default.nix:8]: I don't think we should have a single package depending on cuda arch settings like this.  We already have e.g. config.cudaCapabilities for configuring what architectures cuda-enabled packages are built for.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (bcdarwin) [inline pkgs/development/python-modules/mmcv/default.nix:44]: it shouldn't be necessarily for torch/torchvision to be in nativeBuildInputs and propagatedBuildInputs.

**Review by bcdarwin — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:8]: if there is an example default.nix that will use it that you can point me to?

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: fixed.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:8]: I will try to use torch/default.nix as example

File: pkgs/development/python-modules/mmcv/default.nix
Comment (ConnorBaker) [inline pkgs/development/python-modules/mmcv/default.nix:7]: Not that I'd consider this a blocker for this PR, but for future work consider using `cudaPackages` over `cudatoolkit`. `cudaPackages` uses redistributables from NVIDIA and allows for much smaller closures because you choose which libraries you need. It also makes it easier to maintain because we then have greater visibility into what the package is using.

For an example, take a look at https://github.com/NixOS/nixpkgs/blob/7f194c741b122171494cb8852020310fa57a2944/pkgs/development/python-modules/torchvision/default.nix#L17-L40.

I break the dependencies into those needed only at build time, run time, or both (common). Unfortunately, because most packages expect all the CUDA libraries to be in a single directly, we need to use `symlinkJoin` to group them together.

Also, make sure you're specifying the right C/C++ compilers if you're building with CUDA support, like here: https://github.com/NixOS/nixpkgs/pull/218265. (Torchvision doesn't yet do that; I've got an open issue here: https://github.com/NixOS/nixpkgs/issues/221898.)

The compiler provided by `stdenv` isn't necessarily supported by NVCC, and it's possible to get odd symbol errors during linking if the other derivations were built with a different compiler or language standard. That's why it's important to use `backendStdenv.cc`.

Torchvision and Torch will probably have other things you might want to look at. If you have any questions, don't hesitate to reach out to me or the other CUDA maintainers on Matrix: https://matrix.to/#/#cuda:nixos.org.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:42]: I might use torch torch vision, and let the user overwrite it in a flake file.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:7]: hey @ConnorBaker , thanks a lot for the advices and examples. thanks @bcdarwin  for raising this. I am thinking about using torch.cudaCapabilities as well. but I think there is a valid usecase for people to overwrite with torch-bin (doesn't need to compile the torchWithCuda), which doesn't have passthru for cudaCapabilities. would you suggest that I copy the same pattern from torch/default.nix to get cudaCababilities?

as for the error, I am already getting it with opencv4, as it is compiled with gcc 12, and cudatoolkit uses gcc 11. what is the best to deal with needing to compile with cudatoolkit.cc only for the build phase. use stdenv.cc for the derivation.

Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:7]: also. with following nix-shell, the torchvision fails to build with cuda.

```
with import ../nixpkgs {
  config.allowUnfree = true;
 };
let
  mypthon =
    let
      packageOverrides = self: super: {
        # packagesOverride, will affect all packages that uses the resulted packages as deps
        torch = super.torch.override{ magma = magma-cuda; cudaSupport = true; };
        # opencv4 = super.opencv4.override {
        #   enableBlas = false;
        # };
      };
    in
    python3.override { inherit packageOverrides; self = mypthon; };
in
let
  pythonEnv = mypthon.withPackages (ps: with ps; [ torchvision ]);
in
mkShell {
  packages = [ pythonEnv  ];
}
    


 raise RuntimeError(
RuntimeError: The current installed version of g++ (12.2.0) is greater than the maximum required version by CUDA 11.7 (11.5.0). Please make sure to use an adequate version of g++ (>=6.0.0, <=11.5.0).
/nix/store/c3f4jdwzn8fm9lp72m91ffw524bakp6v-stdenv-linux/setup: line 1593: pop_var_context: head of shell_variables not a function context
        
```

Comment (ConnorBaker) [reply pkgs/development/python-modules/mmcv/default.nix:7]: > but I think there is a valid usecase for people who want to overwrite with torch-bin (doesn't require compilation), but doesn't have passthru for cudaCapabilities.

IIRC, newer torch binaries use CUDA redistributables from PyPi. If they haven't moved over to that yet, then their builder is still using bash scripts to manually copy libraries into their binary :(

There's a hardcoded set of supported capabilities for each version of PyTorch. You can find them here: https://github.com/pytorch/pytorch/blob/49444c3e546bf240bed24a101e747422d1f8a0ee/torch/utils/cpp_extension.py#L1751-L1752.

Because it's a binary distributions and not "built" with Nixpkgs, the CUDA capabilities supported by the binary aren't tied to what we specify in our config.

For that reason I'd recommend you use the same logic `torch/default.nix` has. (Those versions are actually sourced from that same file!)

> as for the odd symbol error, I am already getting that with opencv4, as it is compiled with stdenv.cc gcc 12, and cudatoolkit uses gcc 11. I will give it another go.

I have a PR open which should help with that: https://github.com/NixOS/nixpkgs/pull/221370.

> i assume gcc 12.2 is from stdenv

That's right, it is :(

I know that error well. It's from here: https://github.com/pytorch/pytorch/blob/49444c3e546bf240bed24a101e747422d1f8a0ee/torch/utils/cpp_extension.py#L415. (Related discussion on whether checks like that should be removed when included in Nixpkgs: https://github.com/NixOS/nixpkgs/issues/221564#issuecomment-1478486895.)

That's odd that you're getting it though, since the `torchvision` derivation should correctly set the CC/CXX environment variables to the right compiler: https://github.com/ConnorBaker/nixpkgs/blob/f1d538b45955758f7e1b2e73ab4ee3857a710c9a/pkgs/development/python-modules/torchvision/default.nix#L67-L72.

With respect to your nix file can you try doing something like the following instead? Substitute whatever capability is appropriate for your device (though using just one does make it a lot faster to rebuild stuff from source!).

```nix
with import ../nixpkgs {
  config = {
    allowUnfree = true;
    cudaSupport = true;
    cudaCapabilities = [ "8.6" ];
  };
};
```

I think it was choosing the wrong compiler because `cudaSupport` wasn't explicitly set to `true`.

Let me know how that works!

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mmcv/default.nix:54]: This comment does not justify not using pytestCheckHook at all

```suggestion
```

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mmcv/default.nix:65]:
```suggestion
  nativeCheckInputs = [ pytestCheckInputs ];
```

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/mmcv/default.nix:42]: If someone might do that, they can always overwrite the torch input to this package

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mmcv/default.nix:37]:
```suggestion
    repo = "mmcv";
```

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/mmcv/default.nix:30]: That's 3 months ago. If we introduce a new package and it is a trivial update, we should do that right away unless there is a reason not to

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:7]: @ConnorBaker, your MR on opencv fixed my symbol error.

Comment (ConnorBaker) [reply pkgs/development/python-modules/mmcv/default.nix:7]: Yay! Now it just needs to get merged (somewhat difficult given the large number of rebuilds)!

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: @bcdarwin torch is needed at runtime. not sure there is another way to do it without propagatedBuildInputs

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: fixed.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:81]: I try to do ```
  preCheck = ''
    PYTHONPATH=$out/${python.sitePackages}:$PYTHONPATH
  ''

```
which seems to work with a lot of packages with c extension, but it didn't work for me.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: updated.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: what is pytestCheckInputs

File: pkgs/development/python-modules/mmcv/default.nix
Comment (bcdarwin) [inline pkgs/development/python-modules/mmcv/default.nix:101]: These should almost certainly be moved to `nativeCheckInputs`.

**Review by bcdarwin — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SomeoneSerge) [reply pkgs/development/python-modules/mmcv/default.nix]: `pytestCheckHook`

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/mmcv/default.nix:81]: Another common way is to rm the the conflicting source directory but pytestCheckHook does just invoke pytest as you do, it only builds the arguments to it.

Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:81]: had to find that out the hard way. lol

Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: done

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: done

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: done

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: Done

File: pkgs/development/python-modules/mmcv/default.nix
Comment (bcdarwin) [inline pkgs/development/python-modules/mmcv/default.nix:28]: This comment is now out of date.

Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:28]: it was merged in Nixos:Staging not master. so its still relevant

**Comment by bcdarwin** [comment]
Results of `nixpkgs-review` on x86_64-NixOS:

```
2 packages built:
python310Packages.mmcv python310Packages.mmcv.dist
```

**Review by bcdarwin — APPROVED** [review]
LGTM

**Comment by SomeoneSerge** [comment]
> Results of nixpkgs-review on x86_64-NixOS:

Btw, nixpkgs-review has a `--post-result` option that uses `gh` cli to post a comment containing the exact command you used to run it. Makes it easier to reproduce the results

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mmcv/default.nix:103]:
```suggestion

  buildInputs = [ torch ] ++ lib.optionals cudaSupport [ cuda-redist ];
```

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mmcv/default.nix:76]:
```suggestion
  '';

  postPatch = ''
```

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/mmcv/default.nix:28]: It's currently in staging-next https://nixpk.gs/pr-tracker.html?pr=221370

I would suggest to just leave this open until that is merged which will be very soon.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/mmcv/default.nix:61]: that pname matches the repository name is just coincidence and pname should not be substituted wherever possible.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: fixed

File: pkgs/development/python-modules/mmcv/default.nix
Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:28]: #221370 has been merged into master. @SuperSandro2000

Comment (SuperSandro2000) [reply pkgs/development/python-modules/mmcv/default.nix:28]: So you can resolve the comment now?

Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix]: removed.

File: pkgs/development/python-modules/mmcv/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mmcv/default.nix:105]: Why do we need a formatter here?

Comment (kirillrdy) [reply pkgs/development/python-modules/mmcv/default.nix:105]: yapf is listed as runtime dependency 
https://github.com/open-mmlab/mmcv/blob/v1.7.1/requirements/runtime.txt#L7
derivation fails to build without it

Comment (kirillrdy) [reply pkgs/development/python-modules/mmcv/default.nix:105]: I think  once this is merged 
https://github.com/open-mmlab/mmcv/pull/2151/files
we can remove it

Comment (benxiao) [reply pkgs/development/python-modules/mmcv/default.nix:105]: even when pull request is merged, yapf is still needed for runtime given the code.

Comment (kirillrdy) [reply pkgs/development/python-modules/mmcv/default.nix:105]: yeah true, `Config.dump` still uses it

Comment (SuperSandro2000) [reply pkgs/development/python-modules/mmcv/default.nix:105]: Okay, so it is actually used in the code and not to format code. Thats fine.

**Comment by SuperSandro2000** [comment]
@ofborg build python310Packages.mmcv

**Pull request merged**
